### DISCUSSION TITLE
How to import the @Hilo collection!!

### DISCUSSION THREAD

**Discussion by jeremy-holt** [ravendb/ravendb#19880] [General]
I am creating a new database derived from another database.
I have written an app to go through each collection and map it to the new documents in the new database. There is a lot of complex mapping taking place so I can't just do an "Import database".
Since the documents refer to other documents by @id, I am importing the original document with its' old @id - ie I am not creating a new document in the new database.
Most of the collections on the new db have the same name as the collections in the old db.
This is working fine.
HOWEVER, what I have just realized, is that the @Hilos on the new db (obviously) don't take into account the last @id imported.
Consequently, when I add a new document to the new db, it will overwrite an existing doc with the same @id.

Old Db (@Hilo max 600)
ports/1-A
.....
ports/500-A 

NewDb (@Hilo max 32)
ports/1-A (imported)
....
ports/500-A (imported)
Next save in newDB is ports/1-A (which overwrites the existing ports/1-A).

Short of manually updating each @HiLo in the newDb, is there any way of bringing in the old numbers prorgramatically or writing directly to the @Hilo files and changing the Max?

**Comment by georgiosd** [comment]
Hey @jeremy-holt - you can certainly seed the hilo values yourself like so: https://ravendb.net/docs/article-page/6.2/csharp/client-api/document-identifiers/working-with-document-identifiers#using-operations

**Reply by ayende** [reply]
You are correct, but note that you are talking about _identities_, while @jeremy-holt is talking about documents.
For this purpose, it is important to remember that `@hilo` documents are just that, documents. And you can modify them using the session API.

That is what I would do in this situation

**Comment by jeremy-holt** [comment]
I did try using 'SeedIdentityForOperation' but couldn't get it to work.
What I was doing is (pseudo code)
```
maxId = session.Query<T>().Max 
session.Maintenance.Send(new SendIdentityForOperation(nameof(T).Pluralize, maxId.GetNumberPart()+32)
```

What I would like to be able to do (using Session as Ayende suggests) is
```
hiloCollection = session.GetAllHilos()??
foreach(hilo in hiloCollection){
  hilo.Max = get-my-new-seed-number
}

```
but I can't see how to get the @hilo collection nor how to save it again (await StoreAsync(hilo)? )

**Reply by ayende** [reply]
You can do this:

```
using (var session = store.OpenSession())
{
    var hilos = session.Advanced.LoadStartingWith<dynamic>("Raven/Hilo/",pageSize: 100);
    foreach (var hilo in hilos)
    {
        hilo.Max += 1024;
    }
    session.SaveChanges();
}
```

**Reply by jeremy-holt** [reply]
Unfortunately that updates the hilos in the source database, while I'm trying to copy them from the source to the destination database.

```public class UpdateHilos(IDocumentStore ambBrokerageDocumentStore, IDocumentStore logisticsStore) {
    public async Task Run()
    {
        using (var sourceSession = ambBrokerageDocumentStore.OpenSession())
        using (var destinationSession = logisticsStore.OpenSession())
        {
            {
                var sourceHilos = sourceSession.Advanced.LoadStartingWith<dynamic>("Raven/Hilo/", pageSize: 100); // Get all @Hilo from the source database
                var destinationHilos = destinationSession.Advanced.LoadStartingWith<dynamic>("Raven/Hilo/", pageSize: 100);
                // Remove all existing @Hilo in destinationSession - how to do this?
                // Add/append the sourceHilos to the destinationHilos

                foreach (var sourceHilo in sourceHilos)
                {
                    Enumerable.Append(destinationHilos, sourceHilo); // doesn't work
                }
                
                destinationSession.SaveChanges();
            }
        }

        await Task.CompletedTask;
    }
}```

**Reply by ppekrol** [reply]
Have you tried doing sth like:

```csharp
                    foreach (var sourceHilo in sourceHilos)
                    {
                        string id = sourceHilo.Id;
                        destinationSession.Store(sourceHilo, id);
                    }

                    destinationSession.SaveChanges();
```

**Reply by jeremy-holt** [reply]
I tried that but am getting
```
System.FormatException: The input string 'Raven/Hilo/AppUserPasswords' was not in a correct format.
   at System.Number.ThrowFormatException[TChar](ReadOnlySpan`1 value)
   at System.String.System.IConvertible.ToInt32(IFormatProvider provider)
   at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
   at Newtonsoft.Json.Utilities.ConvertUtils.TryConvertInternal(Object initialValue, CultureInfo culture, Type targetType, Object& value)
   at Newtonsoft.Json.Utilities.ConvertUtils.Convert(Object initialValue, CultureInfo culture, Type targetType)
   at Newtonsoft.Json.Linq.JValue.JValueDynamicProxy.TryConvert(JValue instance, ConvertBinder binder, Object& result)
   at CallSite.Target(Closure, CallSite, Object)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at MappingApp.Mappings.UpdateHilos.Run() in C:\Visual Studio\AmbLogisticsMapper\MappingApp\Mappings\UpdateHilos.cs:line 20
   at Program.<Main>$(String[] args) in C:\Visual Studio\AmbLogisticsMapper\MappingApp\Program.cs:line 125
   at Program.<Main>(String[] args)
   ```

**Reply by ppekrol** [reply]
It must be sth related to your configuration probably? It looks like you are trying to convert string 'Raven/Hilo/AppUserPasswords' to int32 somewhere?

**Comment by jeremy-holt** [comment]
I've also tried this, but when adding a new document, it always uses ports/1-A overwriting the original document

` var sourceSession = ambBrokerageDocumentStore.OpenSession();
        var destinationSession = logisticsStore.OpenSession();
        var query = sourceSession.Query<Port>();
        
        var allDocuments = sourceSession.Advanced.Stream(query);

        while (allDocuments.MoveNext())
        {
            var current = allDocuments.Current;
            var port = current!.Document;
            
            var destinationPort = new Amberwood.AmbLogistics.Server.Models.Dictionaries.Port
            {
                Id = port.Id,
                Name = port.Name,
                Description = port.Description,
                CompanyId = port.CompanyId,
                Country = port.Country
            };
            destinationSession.Store(destinationPort, port.Id);
        }
        
        destinationSession.SaveChanges();`

**Reply by ppekrol** [reply]
Not sure I follow what is the issue here? You are explicitly creating a new destinationPort with given Id, can you explain in more details what are you trying to do and achieve?

**Reply by jeremy-holt** [reply]
I am creating a new database from an existing one, where I am transforming the data in the process. However, I need to preserve the original ID's when creating the new database. This works fine until I add a new document to the new database. 
When I add a document to the new database, because the @hilos were not updated at time of the import/transformation, the new ID's overwrite the old ones.
So what I was trying to do was import the old @hilos collection into the new database.

In any event, I spent the whole day today, rewriting the mapping application, so that it records the old and new ID, and then goes through every document replacing the reference to the old ids with the new one. So, actually I've resolved the problem by doing what I really didn't want to do!!

Just to further explain....
Say I have a document like this in the old db
```
{
Id: contracts/1-A,
UserIds: [users/1-A, users/2-A, user/3-A]
}
```

First of all I import the Users collection using the original Ids, so when I import the Contracts collection the UserIds remain the same. There are many documents with lookups to other collections, and documents related to each other by ID alone. At the time of import the @hilo collection is not updated, because Raven was given the Id to use.
Theoretically this works fine.
However, when I add a new user in the new database, since there is no @hilo entry for Users, Raven gives it an id of users/1-A, thus overwriting the existing users/1-A. Of course what I wanted is that it would give it an id of (say) users/32-A.

Anyway, like I said earlier, I rewrote the app, so that now I ADD the new user at time of importing it, and have a have a seperate document which store the old ID and the new ID. 
I then go through every (say) Contract, looking up the old UserIds from the mapping table and replacing them with the new Ids.

All the same - thanks so much for taking the time to try to help me

Jeremy

**Comment by ayende** [comment]
You can use this code, more or less:

```
 public async Task Run()
    {
        using (var destinationSession = logisticsStore.OpenSession())
	{
		var destinationHilos = destinationSession.Advanced.LoadStartingWith<dynamic>("Raven/Hilo/", pageSize: 100);
         	destinationHilos.ForEach(destinationSession.Delete);
		destinationSession.SaveChanges();
	}
        using (var destinationSession = logisticsStore.OpenSession())
        using (var sourceSession = ambBrokerageDocumentStore.OpenSession())
        {
            {
                var sourceHilos = sourceSession.Advanced.LoadStartingWith<dynamic>("Raven/Hilo/", pageSize: 100); // Get all @Hilo from the source database
         
                foreach (var sourceHilo in sourceHilos)
                {
        		destinationSession.Store(sourceHilo);
			
                }
                
                destinationSession.SaveChanges();
            }
        }

        await Task.CompletedTask;
    }
```
